packages/uqi/README.md

```diff
@@ -0,0 +1,90 @@
+# uqi, a unified query interface
+
+_pronounced: yoo-kee_
+
+## What is uqi?
+
+It is a query interface with pluggable adapters that makes it simple to query different data sources from a single interface.
+
+## Examples
+
+### Querying a CSV file
+
+```typescript
+import { uqi } from '@open-truss/uqi'
+const source = uqi.csv({ path: 'path/to/folder' })
+for await (const { row, metadata } of source.query('SELECT first_name FROM users.csv')) {
+  console.log({ metadata })
+  console.log({ first_name: row.first_name })
+}
+```
+
+### Querying a JSON file
+
+```typescript
+import { uqi } from '@open-truss/uqi'
+const source = uqi.json({ path: 'path/to/users.json', queryLanguage: 'jq' })
+for await (const { row, metadata } of source.query('.[] | select(.first_name | startswith("J")) | .first_name')) {
+  console.log({ metadata })
+  console.log({ first_name: row.first_name })
+}
+```
+
+### Querying a MySQL database
+
+```typescript
+import { uqi } from '@open-truss/uqi'
+const source = uqi.mysql({
+  hostname: process.env.MYSQL_HOSTNAME,
+  username: process.env.MYSQL_USERNAME,
+  password: process.env.MYSQL_PASSWORD,
+  database: process.env.MYSQL_DATABASE,
+})
+for await (const { row, metadata } of source.query('SELECT first_name FROM users')) {
+  console.log({ metadata })
+  console.log({ first_name: row.first_name })
+}
+```
+
+### Querying a Trino database
+
+```typescript
+import { uqi } from '@open-truss/uqi'
+const source = uqi.trino({
+  hostname: process.env.TRINO_HOSTNAME,
+  username: process.env.TRINO_USERNAME,
+  password: process.env.TRINO_PASSWORD,
+})
+for await (const { row, metadata } of source.query('SELECT first_name FROM foo.bar.users')) {
+  console.log({ metadata })
+  console.log({ first_name: row.first_name })
+}
+```
+
+### Querying a Kusto database
+
+```typescript
+import { uqi } from '@open-truss/uqi'
+const source = uqi.kusto({
+  hostname: process.env.KUSTO_HOSTNAME,
+  username: process.env.KUSTO_USERNAME,
+  password: process.env.KUSTO_PASSWORD,
+})
+for await (const { row, metadata } of source.query('foo.bar.users | project first_name')) {
+  console.log({ metadata })
+  console.log({ first_name: row.first_name })
+}
+```
+
+### Querying a Rest API
+
+```typescript
+import { uqi } from '@open-truss/uqi'
+const source = uqi.rest({
+  url: 'http://localhost:3000',
+})
+for await (const { row, metadata } of source.query('GET /users')) {
```

Comment:
REST and JSON are interesting cases because they might not be nicely enumerable like the rest -- a JSON object could be an object and not an array, a REST endpoint might return a single entity.

Do you think there's another single-entity API? Or, would a `for` loop not be used? Or, would uqi wrap non-array results to be an array?

I think my favorite is to just not use a loop.